File: remove_duplicates_array2.cpp
//remove duplicates from sorted array 2
class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        if(nums.size() == 0 || nums.empty()) return 0;

        int slow = 0;
        int fast = 1;
        bool flag = false;

        while(fast < nums.size()){
            if(nums[slow] == nums[fast]) {
                if(!flag){
                    flag = true;
                    slow++;
                    nums[slow] = nums[fast];
                    fast++;
                }else{
                    while(fast < nums.size() && nums[fast] == nums[fast - 1]){
                        fast++;
                    }
                  }
             }else{
                flag = false;
                slow++;
                nums[slow] = nums[fast];
                fast++;
            }
        }
       return slow + 1;
    }
};

//Time Complexity: O(n)
//Space Complexity: O(1)

File: removeduplicatesarray2.java
//Time complexity O(N)
//Space complexity O(1)
//Ran on leetcode and accepted
class Solution {
    public int removeDuplicates(int[] nums) {
        //Initialize 4 pointers
        //Previous
        int prev = 0;
        //Slow pointer
        int slow = 1;
        //Fast pointer
        int fast = 1;
        //flag
        int flag = 0;
        //As long as fast is lesser than the length of the array be in while loop
        while (fast<nums.length)
        {
            //If the value at fast pointer is equal to value at previous pointer 
            if(nums[fast] == nums[prev])
            {
                //Check if flag is 0
                if(flag == 0)
                {
                    //Store the value at fast pointer to the position of slow pointer
                    nums[slow] = nums[fast];
                    //Increment fast , slow and set flag to 1
                    fast++;
                    slow++;
                    flag=1;
                }
                else 
                {
                    //Else only increment fast pointer
                   fast++;
                }
            }
            //else if value at previous pointer and fast pointer is not same
            else 
            {
                //make the index of fast pointer as prev pointer
                prev = fast;
                //set flag to 0
                flag = 0;
                //store value at fast pointer to the position of slow pointer
                nums[slow] = nums[fast];
                //Increment fast annd slow pointers
                fast++;
                slow++;     
            }
        }
        //Return slow pointer as 
        return slow;
    }
}